### ISSUE TITLE
Insertion of coordinates in degree or hours

### ISSUE THREAD

**Issue by giordano** [JuliaAstro/SkyCoords.jl#6]
Currently, coordinates can be inserted only in radians.  What about using an external package which defines physical units in order to allow users to insert coordinates in degree or hours?

If this idea looks good, I'd suggest using [`Unitful.jl`](https://github.com/ajkeller34/Unitful.jl), which has both a simple interface a good performance.  Last time I checked it, [`SIUnits.jl`](https://github.com/keno/SIUnits.jl) had not so good performance

**Comment by kbarbary** [comment]
Well, you *can* input coordinates in degrees with, e.g., `ICRSCoords(deg2rad(ra), deg2rad(dec))`. One could define similar functions `hms2rad(h, m, s)` and `dms2rad([sign], d, m, s)` to handle sexigesmal.

Basically, I'm not yet convinced the added complication of something like Unitful is worth it just for dealing with angles.

**Comment by giordano** [comment]
Good points.  Feel free to close the ticket.

**Comment by mileslucas** [comment]
I've been thinking about a solution to this recently. I think it is beneficial to allow as flexible an input as possible, since there are _so many_ ways to represent coordinates. The other piece of the problem I want to solve is trying to avoid converting input coordinates into an intermediate type.  For example, if a user inputs `c1 = ICRSCoords("5: 3: 56", "23: 4: 11.0")` it would be a better user experience to see
```julia
ICRSCoord(ra="5:3:56", dec="23:4:11.0")
```
instead of 
```julia
ICRSCoord(ra=1.326, dec=0.403)
```
for the returned struct.

This means we probably need to parametrize the coordinates (hopefully avoid issues with the generated FK5 functions). Under the hood, we'd just have to give a little more care to the types in the low-level calculations. That being said, UnitfulAngles makes using `u"rad"` and `u"degree"` really easy; perhaps the low-level calculations just use those? (Would require some benchmarking).

One implementation I can think of would basically add an extra definition like
```julia
lat, lon = latlon(CoordType)

latlon(Coord{<:Number}) # assumes radians
latlon(Coord{<:String}) # parse with str2rad
latlon(Coord{<:Unitlike}) # convert to u"rad" or u"deg" and proceed using UnitfulAngles
```

Thoughts?

**Comment by kbarbary** [comment]
I think putting this stuff into the type system just for the sake of the output of `show` could be more complexity than it is worth. The types are already somewhat complicated.

An alternative that goes some of the way: One could at least give the user control over the `show` output format with a global setting.

**Comment by mileslucas** [comment]
I wanted to show some quick dabbling I did:

If I make something like 
```julia
using Unitful: Quantity, NoDims
const AngleLike = Union{<:AbstractFloat, <:Quantity{<:AbstractFloat, NoDims}
```
and for a quick example I changed

```julia
struct ICRSCoords{T<:AbstractFloat}
# into
struct ICRSCoords{T<:AngleLike}
```

Then I did
```julia
julia> using Unitful: °

julia> c1 = ICRSCoords(0°, 20°)
ICRSCoords{Quantity{Float64,NoDims,Unitful.FreeUnits{(°,),NoDims,nothing}}}(0.0°, 20.0°)

julia> c2 = ans |> GalCoords
GalCoords{Float64}(1.8636665443715206, -0.7199988633872983)

julia> separation(c1, c2)
2.7755575615628914e-17
```

This means that, first off, all of our math Just Works^TM with units. This AngleLike thing let's us say that the user can just give a float and it will be treated as a radian (a good default). Otherwise, if they give degrees, they automatically coerce to floating point versions and can be treated normally.

## Paths forward
1. Always units
I kind of like this idea because it is uber-consistent, but basically we'd convert any floats (assumed to be radian) to `Unitful.rad`. I do realize, though, that an average user may be peeved to have to do extra work when the default trig functions expect floats to be radians. 

2. Partial units
This would be an implementation like I've done above. It would add a dependency on `Unitful` (I don't think that's a problem). Although, we can bypass the need for unitful if we can rewrite the code to not require coordinate inputs to be floats. This means we could do `T<:Number` and get both floats and `Unitful.Quantity`.

Extra Note: UnitfulAngles
I checked out if there is anything super-useful here. The only thing I saw was a nice converter from angles to times (like `rad -> "hh:mm:ss.ss"`). I don't really see the benefit here, since we'll have to have our own parsing code to differentiate string hour angles vs. string degrees. 

----

I'd appreciate any inputs on this matter. I like approach 2, although 1 is my personal favorite (but I can concede its weakness).

**Comment by mileslucas** [comment]
A few extra notes: the code I tested above is all still type stable, and benchmark timing separation and conversion was on order of ~10ns slower.
```julia
julia> using BenchmarkTools

julia> c1 = ICRSCoords(0, 0)
ICRSCoords{Float64}(0.0, 0.0)

julia> @btime c1 |> GalCoords
  263.725 ns (2 allocations: 64 bytes)
GalCoords{Float64}(1.6814027872278692, -1.0504884034813007)

julia> c2 = ICRSCoords(0°, 20°)
ICRSCoords{Float64}(0.0°, 20.0°)

julia> @btime c2 |> GalCoords
  306.020 ns (2 allocations: 64 bytes)
GalCoords{Float64}(1.8636665443715206, -0.7199988633872983)
```
```julia
julia> @btime separation(c1, c1)
  43.639 ns (1 allocation: 16 bytes)
0.0

julia> @btime separation(c2, c2)
  104.464 ns (1 allocation: 16 bytes)
0.0
```

**Comment by helgee** [comment]
I want to share my experience with an "always units" approach which I briefly tried a few years ago. At that point the essential libraries like DifferentialEquations.jl still had to figure out how to be fully generic but that should be sorted out now.
The biggest headache was dealing with constants. As a consequence, constants are implemented as zero-argument functions in AstroBase.jl, e.g.
```julia
const AU = 1.49597870700e8 # km

astronomical_unit(::Type{Float64}) = AU
astronomical_unit() = astronomical_unit(Float64)
```
so that this could be overloaded by a hypothetical UnitfulAstroBase.jl package like this
```julia
using Unitful

astronomical_unit(::Type{<:Quantity}) = astronomical_unit(Float64) * u"km"
```

A function downstream would now use the constant like so
```julia
function foobar(x::T) where T
    au = astronomical_unit(T)
    return x * au
end
```

This works well for standardized constants but I noticed that the algorithms in the literature (in astrodynamics) often contain magical numbers and providing units for them becomes really painful. In the end, I filed a few bug reports, did not go ahead with "always units", and decided to defer the final decision to "Future Helge" who so far has not shown up yet 😜

*TL;DR:* Making the code more generic so that it works with units is probably a good idea. Baking units in probably not.

**Comment by mileslucas** [comment]
Well I could counter argue that we have very few expected units to be used, as well as our current constants being known (they’re all in degrees). 

I think the other big downside of having always units is that if somebody is gun-ho on having liquid metal performance they’d be less than happy. 

Given all that I think just having more generic code is the ultimate better choice for the package, and I think we can implement it fairly simply through the type system.

**Comment by mileslucas** [comment]
With the creation of AstroAngles.jl we don't need to worry about supporting too many different inputs within SkyCoords anymore. Any strings will be parsed by AstroAngles, any Reals will be assumed as radians, and in the future we can add support for Unitful.°. By splitting the parsing out (as suggested by @helgee) many more packages can benefit, and this package can focus more on the coordinate system types and conversions.
